src/core/Serialize.ml

```diff
@@ -0,0 +1,729 @@
+open Basis
+open Bwd
+open Cubical
+open CodeUnit
+
+module S = Syntax
+module D = Domain
+module J = Ezjsonm
+
+(* Basic JSON Helpers *)
+let json_of_pair x y = `A [x; y]
+
+let json_to_pair json_to_a json_to_b =
+  function
+  | `A [j_a; j_b] ->
+    let a = json_to_a j_a in
+    let b = json_to_b j_b in
+    (a,b)
+  | j -> J.parse_error j "json_to_pair"
+
+let json_of_int n = `String (Int.to_string n)
+
+let json_to_int : J.value -> int =
+  function
+  | `String n -> int_of_string n
+  | j -> J.parse_error j "json_to_int"
+
+let json_of_alist json_of_a json_of_b xs : J.value list =
+  List.map (fun (a, b) -> json_of_pair (json_of_a a) (json_of_b b)) xs
+
+let json_to_alist json_to_a json_to_b j_alist : ('a * 'b) list =
+  List.map (json_to_pair json_to_a json_to_b) j_alist
+
+let json_of_bwd json_of_el bwd : J.value =
+  `A (Bwd.to_list @@ Bwd.map json_of_el bwd)
+
+let json_to_bwd json_to_el : J.value -> 'a bwd =
+  function
+  | `A xs -> Bwd.map json_to_el @@ Bwd.from_list xs
+  | j -> J.parse_error j "json_to_bwd"
+
+let labeled lbl v = `A (`String lbl :: v)
+
+(* Identitifers *)
+let label_of_ident : Ident.t -> string =
+  function
+  | `Anon -> "anon"
+  | `User parts -> "user/" ^ String.concat "." parts
```

Comment:
Since we are using JSON, can we just use `{user:"blah"}` instead of `"user/blah"`?

Reply:
These are used as keys for JSON objects, so they have to be strings. I suppose an alist would also work though!